### DISCUSSION TITLE
LiDAR configurations

### DISCUSSION THREAD

**Discussion by thomasTUK** [microsoft/AirSim#4406] [Support Q&A]
Hello there,

I would like to set the setting to a specific sensor. For example the Velodyne Puck

![grafik](https://user-images.githubusercontent.com/79101120/157906024-2bd239cf-3f8e-4e61-ba68-a0cbc851a05a.png)

How do I do that correctly? I would do it like this:

                "SensorType": 6,
                "Enabled" : true,
                "NumberOfChannels": 16,
                "Range": 100,
                "PointsPerSecond": 300000,
                "RotationsPerSecond": 5,
                "HorizontalFOVStart": 0,
                "HorizontalFOVEnd": 0,
                "VerticalFOVUpper": 15,
                "VerticalFOVLower": -15,
 
The vertical resolution is FOV/Channels-1 = 2. But what about thr horizontal angular resolution? As far as I know is that it responds to the rotation rate like 5Hz = 0,1° and 20Hz = 0,4° (values between correspond). 

Do you guys agree?

**Comment by thomasTUK** [comment]
In UnrealLidarSensor.cpp there is following equation:

    // calculate number of points needed for each laser/channel
    const uint32 points_to_scan_with_one_laser = FMath::RoundHalfFromZero(total_points_to_scan / float(number_of_lasers));

and

    // calculate needed angle/distance between each point
    const float angle_distance_of_tick = params.horizontal_rotation_frequency * 360.0f * delta_time;
    const float angle_distance_of_laser_measure = angle_distance_of_tick / points_to_scan_with_one_laser;


Can someone explain to me, what delta_time is?
Many thanks!

**Reply by zimmy87** [reply]
delta_time is derived from the lidar's update frequency, which is currently set to 10 Hz [here](https://github.com/microsoft/AirSim/blob/master/AirLib/include/sensors/lidar/LidarSimpleParams.hpp#L43), so delta_time should be around ~0.1s (the real value will be slightly different to account for the physics update frequency). Right now, this value is only set inside of AirLib, so changing the value will require a source change, either to manually set the value or to add support for reading it from a setting inside AirSimSettings.

Right now, the lidar is implemented in a way that groups consecutive laser scans together. The [getPointCloud](https://github.com/microsoft/AirSim/blob/master/Unreal/Plugins/AirSim/Source/UnrealSensors/UnrealLidarSensor.cpp#L43) implementation simulates multiple laser scans that would have taken place over the time period with length delta_time that elapses between each call of [updateOutput](https://github.com/microsoft/AirSim/blob/master/AirLib/include/sensors/lidar/LidarSimple.hpp#L76).

**Reply by thomasTUK** [reply]
Hi @zimmy87 so i have change in the source file (UnrealEngine\EnvironmentName\Plugins\AirSim\Source\AirLib\include\sensors\lidar\LidarSimpleParams.hpp) the value of [real T update_frequency](https://github.com/microsoft/AirSim/blob/master/AirLib/include/sensors/lidar/LidarSimpleParams.hpp#L43) to 5 or 20 Hz, but I got with the same settings.json the exact number of point (29377).

So, in my case, delta_time is not derived from the update frequency - why is that? Looking forward to your anwser :)

**Comment by VANNVisal** [comment]
Dear Mr. I have a problem about the run Drone_lidar.py  we have a problem in file settings.json i don't know a bout this problem so you can help me pleas.And I don't see setting.json or path that store setting.json that connect with Drone_lidar.py and about process for build and run python file. 

this (venv_python39) PS C:\Users\BNC\Documents\ITC-Internship\CodeAirsim>  & 'c:\Users\BNC\anaconda3\envs\venv_python39\python.exe' 'c:\Users\BNC\.vscode\extensions\ms-python.debugpy-2024.10.0-win32-x64\bundled\libs\debugpy\adapter/../..\debugpy\launcher' '56334' '--' 'c:\Users\BNC\Documents\ITC-Internship\CodeAirsim\drone_lidar.py' 
Connected!
Client Ver:1 (Min Req: 1), Server Ver:1 (Min Req: 1)

Connected to AirSim and API control enabled.
Arming the drone...
['SimpleFlight']
Traceback (most recent call last):
  File "c:\Users\BNC\Documents\ITC-Internship\CodeAirsim\drone_lidar.py", line 129, in <module>
    lidarTest.execute()
  File "c:\Users\BNC\Documents\ITC-Internship\CodeAirsim\drone_lidar.py", line 19, in execute
    print(self.client.getLidarData("lidar_2"))
  File "c:\Users\BNC\anaconda3\envs\venv_python39\lib\site-packages\airsim\client.py", line 906, in getLidarData

this my code 👍Drone_lidar.py
import airsim
import time
import argparse
import pprint
import numpy
import keyboard  # Import the keyboard module for handling key presses

class LidarTest:
    def __init__(self):
        # Connect to the AirSim simulator
        self.client = airsim.MultirotorClient()
        self.client.confirmConnection()
        self.client.enableApiControl(True)
        print("Connected to AirSim and API control enabled.")

    def execute(self):
        print("Arming the drone...")
        print(self.client.listVehicles())
        print(self.client.getLidarData("lidar_2"))
        self.client.armDisarm(True)
        
        state = self.client.getMultirotorState()
        print("Drone armed. Current state:")
        print(pprint.pformat(state))

        print('Press "t" to takeoff, "q" to land, "w" to move forward, "s" to move backward, "a" to move left, "d" to move right, "h" to hover, "l" to get Lidar readings, "x" to stop.')

        lidar_name = "lidar_2"  # Ensure this matches the name in your settings.json

        while True:
            if keyboard.is_pressed('t'):  # Takeoff
                print("Taking off...")
                self.client.takeoffAsync().join()
                print("Takeoff complete.")
                time.sleep(1)
                
            if keyboard.is_pressed('q'):  # Land
                print("Landing...")
                self.client.landAsync().join()
                print("Landing complete.")
                time.sleep(1)
                
            if keyboard.is_pressed('w'):  # Move forward
                print("Moving forward...")
                self.client.moveByVelocityAsync(10, 0, 0, duration=5).join()
                print("Move forward complete.")
                time.sleep(1)
                
            if keyboard.is_pressed('s'):  # Move backward
                print("Moving backward...")
                self.client.moveByVelocityAsync(-10, 0, 0, duration=5).join()
                print("Move backward complete.")
                time.sleep(1)
                
            if keyboard.is_pressed('a'):  # Move left
                print("Moving left...")
                self.client.moveByVelocityAsync(0, -10, 0, duration=5).join()
                print("Move left complete.")
                time.sleep(1)
                
            if keyboard.is_pressed('d'):  # Move right
                print("Moving right...")
                self.client.moveByVelocityAsync(0, 10, 0, duration=5).join()
                print("Move right complete.")
                time.sleep(1)
                
            if keyboard.is_pressed('h'):  # Hover
                print("Hovering...")
                self.client.hoverAsync().join()
                print("Hover complete.")
                time.sl
            if keyboard.is_pressed('l'):  # Get Lidar readings
               
                try:
                    print(f"Retrieving Lidar data from {lidar_name}...")
                    for i in range(1, 5):
                     
                        lidarData = self.client.getLidarData()
                        
                        if len(lidarData.point_cloud) == 0:
                            print(f"\tReading {i}: No points received from Lidar data")
                        else:
                            points = self.parse_lidarData(lidarData)
                            print(f"\tReading {i}:")
                            print(f"\t\tTime Stamp: {lidarData.time_stamp}")
                            print(f"\t\tNumber of Points: {len(points)}")
                            print(f"\t\tLidar Position: {pprint.pformat(lidarData.pose.position)}")
                            print(f"\t\tLidar Orientation: {pprint.pformat(lidarData.pose.orientation)}")
                            # Print first few points for inspection
                            print("\t\tFirst few Lidar points:")
                            for point in points[:5]:  # Print only the first 5 points for brevity
                                print(f"\t\t\t{point}")
                        time.sleep(5)
                except Exception as e:
                    print(f"Error retrieving Lidar data: {e}")
                
            if keyboard.is_pressed('x'):  # Stop and exit
                print("Stopping and exiting...")
                break
        
        self.stop()

    def parse_lidarData(self, data):
        # Reshape array of floats to array of [X, Y, Z]
        if not data.point_cloud:
            return numpy.array([]).reshape(0, 3)
        
        points = numpy.array(data.point_cloud, dtype=numpy.dtype('f4'))
        points = numpy.reshape(points, (int(points.shape[0] / 3), 3))
        return points

    def write_lidarData_to_disk(self, points):
        # TODO: Implement saving Lidar data to disk
        print("Saving Lidar data to disk is not yet implemented.")

    def stop(self):
        print('Stopping...')
        self.client.armDisarm(False)
        self.client.reset()
        self.client.enableApiControl(False)
        print("Drone disarmed and reset to original state.")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Control a drone and retrieve Lidar data.")
    parser.add_argument('-save-to-disk', type=bool, help="Save Lidar data to disk", default=False)
    args = parser.parse_args()

    lidarTest = LidarTest()
    lidarTest.execute()
 Thank you!
